### DISCUSSION TITLE
Graph direction

### DISCUSSION THREAD

**Discussion by MaxDragonheart** [networkx/networkx#5469] [Q&A]
I'm working on a road network and I think I have an incorrect approach to direction of travel.
For problem related to a conflict between networkx and scipy I can't use momepy, so I've developed a workaround to read my road from a GeoDataFrame:
```
def make_directional_graph(
        input_data: geopandas.geodataframe.GeoDataFrame,
        input_geometry_column: str = 'geometry',
) -> networkx.classes.digraph.DiGraph:
    """Make directional graph from linestring vector.
    Args:
        input_data: geopandas.geodataframe.GeoDataFrame
        input_geometry_column: str
    Returns:
        nx.classes.graph.DiGraph
    """
    # Make graph
    graph = nx.DiGraph()
    for _index, row in line_data.iterrows():
        line_geometry = row[input_geometry_column]
        line_coords = list(line_geometry.coords)
        first_point = line_coords[0]
        last_point = line_coords[-1]
        direction = row.direction

        if direction == 0:
            # Edge bidirectional
            graph.add_edge(first_point, last_point)
            graph.add_edge(last_point, first_point)
        elif direction == 1:
            # Edge clockwise
            graph.add_edge(first_point, last_point)
        else:
            # Edge anti-clockwise
            graph.add_edge(last_point, first_point)

    return graph
```
Maybe the function can be useful for someone else.
[Here](https://drive.google.com/file/d/1g7dA4A0sEvCgGmPEBjtLlECaywv-BFGp/view?usp=sharing) there are my sample roads. When I read the vector with the function above I see this:

![Screenshot from 2022-04-05 09-24-18](https://user-images.githubusercontent.com/18444902/161700851-0dd3c315-b421-4a2b-ba9c-8a7e7f960959.png)

According to `direction` field in the attribute table I expect that the "*ring*" has the same direction(clockwise), but only the half-left is clockwise. So I realize that I'm not reading the `direction` field but the way it was drawn every edge. 
For the left-half I drew the edges using a direction, for the right-half another. I don't think that is a problem of networkx because I see the same thing on QGIS.
It is not clear for me how I can read really the edge direction from the attribute table.

**Comment by rossbar** [comment]
Are you sure this isn't a visualization problem? Without the code used to create the plot (including the node positions) it's hard to say. Note that if you didn't explicitly set the node positions when plotting, then chances are they were chosen using `spring_layout`, which is the default layout function for `nx.draw`, and will almost certainly result in different node positions from what you expect.

Generally this seems like it might be a better fit for the GeoPandas issue tracker.

**Reply by MaxDragonheart** [reply]
@rossbar I found [this article](https://sustainability-gis.readthedocs.io/en/latest/lessons/L2/spatial_network_analysis.html) that is quite similar to mine approach. Below my code, I'm using the same sample roads from my previous post.

With GeoPandas I read the shapefile, he contains only three informations: fid, direction and geometry. Now I have the GeoDataFrame named `roads` from which I extract the nodes:
```
import pandas as pd

from_nodes = []
to_nodes = []

for _index, row in roads.iterrows():
    road_id = row.fid
    geometry =  row.geometry
    coordinates = list(geometry.coords)
    from_node = coordinates[0]
    to_node = coordinates[1]

    from_nodes.append([road_id, from_node])
    to_nodes.append([road_id, to_node])

df_from = pd.DataFrame(from_nodes)
df_from.rename(columns={0: 'road_id', 1: 'coordinates'}, inplace=True)

df_to = pd.DataFrame(to_nodes)
df_to.rename(columns={0: 'road_id', 1: 'coordinates'}, inplace=True)

nodes_df = pd.concat([df_from, df_to], ignore_index=True)
```
After that I remove the duplicated nodes and assign an ID to all single nodes:
```
node_coordinates = nodes_df['coordinates'].unique()
nodes_unique = pd.DataFrame(node_coordinates)
nodes_unique.reset_index(inplace=True)
nodes_unique.rename(columns={'index': 'node_id', 0: 'coordinates'}, inplace=True)
```
Now I can add node ID and it's coordinates to the single road:
```
road_data_list = []

for _index, row in roads.iterrows():
    road_id = row.fid
    direction = row.direction
    geometry =  row.geometry
    coordinates = list(geometry.coords)
    from_node = coordinates[0]
    to_node = coordinates[1]

    from_node_specification = {}
    to_node_specification = {}
    for _index2, row2 in nodes_unique.iterrows():
        node_id = row2.node_id
        node_coordinates = row2.coordinates

        if node_coordinates == from_node:
            from_node_data = {'node_id': node_id, 'coordinates': from_node}
            from_node_specification.update(from_node_data)
        elif node_coordinates == to_node:
            to_node_data = {'node_id': node_id, 'coordinates': to_node}
            to_node_specification.update(to_node_data)
        else:
            pass

    road_data = [road_id, geometry, from_node_specification, to_node_specification, direction]
    road_data_list.append(road_data)

roads_df = pd.DataFrame(road_data_list)
roads_df.rename(columns={0: 'road_id', 1: 'geometry', 2: 'from_node', 3: 'to_node', 4: 'direction'}, inplace=True)
```
Finally I build a graph using `direction` parameter to establish the travel direction:
```
import networkx as nx

edge_list = []

for _index, row in roads_df.iterrows():
    direction = row.direction
    from_node_id = row.from_node['node_id']
    to_node_id = row.to_node['node_id']

    if direction == 0: # bidirectional road
        clockwise_direction = [from_node_id, to_node_id]
        anticlockwise_direction = [to_node_id, from_node_id]
        edge_list.append(clockwise_direction)
        edge_list.append(anticlockwise_direction)
    elif direction == 1: # oneway
        clockwise_direction = [from_node_id, to_node_id]
        edge_list.append(clockwise_direction)
    elif direction == -1: # oneway
        anticlockwise_direction = [to_node_id, from_node_id]
        edge_list.append(anticlockwise_direction)
    else:
        pass

graph = nx.MultiDiGraph()
graph.add_edges_from(edge_list)
```
Plotting the graph I see this:
```
from matplotlib import pyplot as plt

fig, ax = plt.subplots(figsize=(10, 10))

nx.draw(
    graph,
    label='Graph',
    with_labels=True
)

plt.legend()

plt.show()
```
![Screenshot from 2022-04-07 09-15-02](https://user-images.githubusercontent.com/18444902/162141658-5d8f35c2-79c4-4a8b-ac60-f8e8c1b687cc.png)
When I gave the direction to the roads I assigned the same oneway direction to:
- 1 -> 3
- 3 -> 4
- 4 -> 5
- 5 -> 2
- 2 -> 1

Looking the graph, the direction for `5 -> 2` and `2 -> 1` is opposite. I don't understand if is only a visualization problem or if my approach is uncorrect. Keep in mind that after I need to find the shortest path and if my approach is uncorrect even the short path it won't be.

**Reply by dschult** [reply]
The notion of clockwise and anti-clockwise does not make sense here... but I think that is just a naming convention for your variables and will not affect anything.  The edges you think exist in `graph` differ from what you think they are.  You say that you assigned the edges `(1 -> 3, 3 -> 4, 4 -> 5, 5 -> 2, 2 -> 1)`.   But the evidence suggests you assigned `(1 -> 3, 3 -> 4, 4 -> 5, 2->5, 1->2, 1->0, 0->1, 5->6, 6->5)`.  You can check all the edges in the graph using `print(graph.edges())`. The order of the nodes in each 2-tuple is the direction of the arrow on the drawing.

**Reply by rossbar** [reply]
My point re: clockwise/counter-clockwise was that it depends on the node positions - for instance, if nodes 3 and 4 from the latest image were reflected about the line defined by `5 -> 1`, the loop (`5-1-3-4`) will then be "clockwise" in the image! Since you're not providing node positions, they are being chosen for you under-the-hood by the `spring_layout`, which will likely give you different (but similar) layouts each time you use `nx.draw`.

I agree with @dschult though, it seems like the root of this particular problem is that graph you're creating likely has different structure that you expect.

**Reply by MaxDragonheart** [reply]
@dschult yes, clockwise and anticlockwise is only a naming convention. So the problem isn't in the method with which I build the graph, but it is in how I populate the column `direction`.  Right?
I use the column `direction` for build the graph:
```
edge_list = []

for _index, row in roads_df.iterrows():
    direction = row.direction
    from_node_id = row.from_node['node_id']
    to_node_id = row.to_node['node_id']

    if direction == 0: # bidirectional road
        clockwise_direction = [from_node_id, to_node_id]
        anticlockwise_direction = [to_node_id, from_node_id]
        edge_list.append(clockwise_direction)
        edge_list.append(anticlockwise_direction)
    elif direction == 1: # oneway
        clockwise_direction = [from_node_id, to_node_id]
        edge_list.append(clockwise_direction)
    elif direction == -1: # oneway
        anticlockwise_direction = [to_node_id, from_node_id]
        edge_list.append(anticlockwise_direction)
    else:
        pass

graph = nx.MultiDiGraph()
graph.add_edges_from(edge_list)
```
If this method is OK I will be able to focus my effort on the method with which I assign the direction; this last case isn't a programming problem but a pure transport engineering problem.
